### PR TITLE
[VFX] Additional Sorting Keys [22.1]

### PR DESCRIPTION
**Pull request by ludovic-theobald**
# **[VFX] Additional Sorting Keys**

### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [ ] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
This PR introduces new criteria to sort the particles of an output context. Prior to this PR, if sorting was enabled, particles were sorted by their distance to the camera. We now propose a total of 5 sorting criteria : 

- By Distance (previously, the only option)
- Youngest in front
- Oldest in front
- By Depth
- Custom
    - the Custom criterion creates a new slot in the Output context, where the user can plug a sorting key which will be used to sort the particles.
SortKeySlot

The sorting criterion can be chosen in the inspector if the **Sort** is set to On or Auto.
SortCriterionMenu
FilteredOut_SortCriterionMenu

Notes: 

- We cannot sort particles from different outputs, meaning that if you have two outputs to a system, one will appear completely in front or completely behind, depending on their sort priority.
- When possible (outputs sharing the same sort criterion), only one global sort is issued for all the outputs. If outputs have different sorting criteria, they will require their own sort pass, incurring an additional cost in performance.
- This [bug](https://fogbugz.unity3d.com/f/cases/1203011/) causing flickering when there are more than 4096 particles is fixed in this PR.

### [JIRA story](https://jira.unity3d.com/browse/GVFXG-128)

This PR required modification of the C++ code.
### Link to C++ PR : https://ono.unity3d.com/unity/unity/pull-request/130921/_/graphics/vfx/feature/new-sortkeys

---
### Testing status
What has been tested : 
- Sorting 1 output (simple quads) with all the sorting criterion
- Sorting up to 4 outputs, with heterogeneous sorting criteria
- Sorting two multimesh outputs
- Only in HDRP (but shouldn't be dependant on the SRP)

What hasn't been tested :
- Different platforms
- Extensive test on the custom sort key slot (e.g. will it break with some inputs?)
- Upgrading from an older version

I plan to add a graphics test for this feature too.

---

### CODE REVIEW

**Comment by VladNeykov** [comment]
Hi, dropping the [WIP Test Doc](https://docs.google.com/document/d/1czRW_PxEv-IYJGP38Cd__iOMGP1SbeZc1AJ2XFiHhXM/edit#) with a few items already that can be looked at.

File: com.unity.visualeffectgraph/Shaders/Sort.compute [hunk 1]
```diff
@@ -290,15 +310,14 @@ void MERGE_PASS(uint3 groupId           : SV_GroupID,
     } while (!done);
 
     // Select left or right: Must handle special case for equality
-    bool select0 = (key0 == key1 && (id & 1)) || (CompareKeys(key0,key1) != reverse);
-    uint value = inputSequence[(select0 ? index0 : index1)].value;
+    bool select0 = (kvp0.key == kvp1.key && kvp0.value == kvp1.value && (id & 1)) || (CompareKVP(kvp0, kvp1) != reverse);
```
Comment (julienf-unity) [inline com.unity.visualeffectgraph/Shaders/Sort.compute:313]: The first condition does not seem relevant anymore. CompareKVP test seems enough

File: com.unity.visualeffectgraph/Shaders/Sort.compute [hunk 2]
```diff
@@ -290,15 +310,14 @@ void MERGE_PASS(uint3 groupId           : SV_GroupID,
...
 #if FINAL_PASS
-    sortedSequence[id] = value;
+    sortedSequence[id] = select0 ? kvp0.value : kvp1.value;
 #else
-    float key = select0 ? key0 : key1;
-    KVP kvp = { key, value };
-    sortedSequence[id] = kvp;
+    if (select0)
+        kvp1 = kvp0;
+    sortedSequence[id] = kvp1;
```
Comment (julienf-unity) [inline com.unity.visualeffectgraph/Shaders/Sort.compute:320]: I would use a ternary operator here

**Review by julienf-unity — COMMENTED** [review]
In progress

File: com.unity.visualeffectgraph/Shaders/Sort.compute (hunk 2 above)
Comment (ludovic-theobald) [reply com.unity.visualeffectgraph/Shaders/Sort.compute:320]: It turns out that the compiler doesn't like user defined types in ternary operator, that's why it was like that in the first place :(

**Comment by ludovic-theobald** [comment]
Guardians from outside of VFX were added automatically due to the reformatting affecting other files, please disregard.

**Review by VladNeykov — APPROVED** [review]
Looking great, no outstanding issues ([test doc](https://docs.google.com/document/d/1czRW_PxEv-IYJGP38Cd__iOMGP1SbeZc1AJ2XFiHhXM/edit#)) 
Thanks for this feature, it's super useful! :)

**Comment by sebastienlagarde** [comment]
any update on this PR?

**Comment by ludovic-theobald** [comment]
> any update on this PR?

C++ side of the PR is currently being reviewed

**Review by Vic-Cooper — APPROVED** [review]
I approve my own changes :D

**Review by julienf-unity — APPROVED** [review]

**Review by PaulDemeulenaere — APPROVED** [review]
Approving but only because I'm following the landing of this PR (see https://github.com/Unity-Technologies/Graphics/pull/6172)

File: TestProjects/VisualEffectGraph_HDRP/Assets/AllTests/VFXTests/GraphicsTests/16_MeshParticles.unity [hunk 3]
```diff
@@ -317,12 +370,115 @@ VisualEffect:
...
+    IncorrectPixelsThreshold: 0.0000038146973
+    UseHDR: 0
+    UseBackBuffer: 0
+    ImageResolution: 0
+    ActiveImageTests: 1
+    ActivePixelTests: 7
+  captureFrameRate: 20
+  fixedTimeStepScale: 1
+  simulateTime: 5.95
+  xrCompatible: 0
```
Comment (PaulDemeulenaere) [inline TestProjects/VisualEffectGraph_HDRP/Assets/AllTests/VFXTests/GraphicsTests/16_MeshParticles.unity:484]: Discussed with @ludovic-theobald in PM, `xrCompatible: 0` is expected 🟢

**Pull request merged**